Arc printing optimisations

tombrazier opened the pull request:
### Description
Printing arcs can max out an AVR processor. These optimisations significantly reduce processors usage by giving the planner a few hints to help it in its work.

The new `planner_hints_t` type could be used by `G5` Bézier curves or even new parameters to `G0/G1` to give the same benefits to other paths with many short segments.

#### Optimisation 1
Printing the following circle takes about 3s on my ATMega2560 machine.
```
G3 X150 Y100 I-50 J0 F10000
```

During this time, `idle()` is rarely called because the CPU is pretty much maxed out (a total of 2.6s of non-idle CPU time is needed just to draw the circle). Consequently, the planner queue depth drops well below 15. 1.6s of this time is spent in `Planner::recalculate_trapezoids()` because every time a new short arc segment is added, the speed calculations have to be propagated all the way back through the entire queue. This can be prevented if the arc code informs the trapezoidal calculation code about the remaining as yet unbuffered arc.

This change takes 1.1s off the total non-idle CPU time, taking it down to 1.5s of CPU time which is a utilisation of about 50%.

This also allows some arcs to be printed faster because the planner can now see further into the future.

#### Optimisation 2
Junction deviation calls `sqrt()` and has some floating point multiplications and divisions which can all be avoided if the radius of curvature is actually known. Then the cornering speed limit squared is simply `acceleration limit * radius`.

On the same 50mm radius circle, this saves about 50ms. This is a further 3% saving on the new value of 1.5s CPU time.

Smaller circles also show savings. The following 10mm radius circle saves about 15ms out of 365ms.
```
G3 X150 Y100 I-10 J0 F10000
```

#### More details
[Update: based on the ideas below, there is now a new type `CodeProfiler` which is enabled when `MAX7219_DEBUG` is enabled. This helps a lot with measuring the time spent in any particular part of the code.]

To test timings, I placed calls to `micros()` before and after the relevant code. `idle()` needed some special treatment because it gets called from everywhere so I created the following class
```
class idletimer {
    static uint32_t totaltime;
    uint32_t starttime;
  public:
    idletimer() : starttime(micros()) {}
    ~idletimer() {totaltime += micros() - starttime;}
    static void reset() {totaltime = 0;}
    static uint32_t gettime() {return totaltime;}
};
```
and instantiated an instance of it at the start of `idle()`. This allowed me to call track all time spent inside `idle()` called from anywhere.

### Requirements

It should be of most benefit to AVR boards but affects all.

### Benefits

Fewer CPU cycles. Less chance of stutter. Longer life for AVR.

### Configurations

Enable `ARC_SUPPORT`.

tombrazier committed:
sqrt() is a lot cheaper than cos() or sin()

tombrazier committed:
Use known arc radius to save time in the junction deviation code

rhapsodyv commented:
Why add a new config? Is there any issue making the new code always enabled?

tombrazier commented:
Technically this is no longer junction deviation. I had a nagging feeling I needed to be up front about that. But then at the time I was thinking junction deviation was something you choose in config, whereas it is actually just the default if you disable classic jerk.

There is no potential downside I am aware of. A slight upside is marginally faster cornering because JD assumes the toolhead is going to cut the corner, giving a very slightly smaller radius than the arc code would.

Perhaps splitting this particular hair is just confusing. I think I will remove the config change.

thisiskeithb commented:
I'm surprised the LERDGEK test hasn't timed out already. It seems stuck downloading.

tombrazier commented:
Okay, I have now found a 1.1 million microsecond reduction in AVR processor usage when printing the 50mm diameter circle mentioned above. Which really does make the other optimisations look a tad anaemic.

@thinkyhead it would be helpful to get some guidance on the best way to implement it. I have just done it via a global variable bodge at the moment to test the idea. Here's the diff: https://github.com/tombrazier/Marlin/commit/b99c13ec5665599471d9859f9aef041c4ac7e148

The change is simple. It tells the trapezoidal calculations how much of the arc is remaining still to print. As a consequence for most of the arc the planner blocks do not need recalculating because they are far from the end of the arc movement. I can now print circles with my planner queue depth remaining at 15 or 16 for the duration.

DerAndere1 commented:
is the `radius` argument also used in G0/G1 moves? otherwise I would revert commit [d23d9f4](https://github.com/MarlinFirmware/Marlin/pull/24366/commits/d23d9f4c290d649c44d6f003440cc552c652b349) and make the `arc_radius` in planner.cpp an OPTARG (maybe depending on existing config option `ARC_SUPPORT` instead of `ARC_ASSISTED_JD`).

tombrazier commented:
> is the arc_radius argument also used in G0/G1 moves? otherwise I would partially revert commit [d23d9f4](https://github.com/MarlinFirmware/Marlin/pull/24366/commits/d23d9f4c290d649c44d6f003440cc552c652b349) and make the arc_radius in planner.cpp an OPTARG (maybe depending on existing config option ARC_SUPPORT instead of ARC_ASSISTED_JD).

I woke up in the middle of the night with exactly the same thought. Will do.

I also think I will now remove the sin=sqrt(1-cos*cos) change. With even an AVR CPU getting bored from being so idle, I think the saving is not worth the slight extra complexity on the code.

tombrazier committed:
Just do it!

tombrazier committed:
Revert "sqrt() is a lot cheaper than cos() or sin()"

This reverts commit 361268bf479937725a650ff09269a0c24be5cdb2.

The CPU time saved does not justify the extra code complexity.

tombrazier commented:
I have removed the sqrt vs sin optimisation and edited the OP and subject so this PR is now about the two more significant optimisations.

thinkyhead commented:
Try your changes with different speed limits on the X and Y axes. Is it possible that a trapezoid really is needed at every junction in the curve, or can it maintain constant feedrates as anticipated?

thinkyhead committed:
Ti

thinkyhead committed:
clean up

tombrazier committed:
Fix recalculate_trapezoids with PAGE block

Also cleaned up the code at the end so it is far more consistent name-wise with the block inside the loop above

tombrazier commented:
@thinkyhead My last commit fixes what I think is a subtle bug. The the first loop in Planner::recalculate_trapezoids() is trying to ensure the last block of code only calculates trapezoids for the right kind of block. But it does not test for and exclude PAGE blocks.

I *also* renamed the `next` and `next_entry_speed` variables in the last block because they will have been copied into `block` and `current_entry_speed` by this point. So this block of code can look just like the code in the loop above. This makes sense because the code block is doing exactly the same thing, except for using MINIMUM_PLANNER_SPEED as the exit speed.

This will make my next commit - with the big optimisation - easier.

It also suggests a cleanup where the two almost identical blocks are merged.

I think the name changes also render the first loop in the function redundant. There is no longer a need to ensure that `next` does not land on a SYNC or PAGE block because we are using `block`, which is guaranteed not to be a SYNC or PAGE block because of this test
https://github.com/MarlinFirmware/Marlin/blob/7535a16f708db2e46bba836aea313d66ea925664/Marlin/src/module/planner.cpp#L1165

I wasn't quite brave enough to delete the first loop, though.

thinkyhead commented on Marlin/src/module/planner.cpp:
```diff
@@ -1143,7 +1143,7 @@ void Planner::recalculate_trapezoids() {
 
     // If not dealing with a sync block, we are done. The last block is not a SYNC block
-    if (!(prev->flag & BLOCK_MASK_SYNC)) break;
+    if (!(prev->flag & BLOCK_MASK_SYNC) && !IS_PAGE(prev)) break;
```
@colinrgodsey — Is this correct, something we missed in #17853 ?

colinrgodsey replied:
ooh yea, this should be added here. originally (iirc) i didn't update this because it was only getting called when buffering segments. but the page support doesnt actually prevent you from using traditional movement blocks, so this can still get called. good catch!

thinkyhead replied:
Thanks for the reply. You never know if there is some subtle reason….

thinkyhead committed:
Bit fields for block flags

tombrazier commented on Marlin/src/module/planner.cpp:
```diff
@@ -1107,7 +1106,7 @@ void Planner::forward_pass() {
 
     // Skip SYNC and page blocks
-    if (!(block->flag & BLOCK_MASK_SYNC) && !IS_PAGE(block)) {
+    if (!block->is_move()) {
```
I think this should be `block->is_move()` - no not.

tombrazier commented on Marlin/src/module/planner.cpp:
```diff
@@ -1142,8 +1141,8 @@ void Planner::recalculate_trapezoids() {
-    // If not dealing with a sync block, we are done. The last block is not a SYNC block
-    if (!(prev->flag & BLOCK_MASK_SYNC) && !IS_PAGE(prev)) break;
+    // If not dealing with a sync or page block, we are done. The last block is not a SYNC or PAGE block.
+    if (!prev->is_move()) break;
```
ditto

tombrazier commented on Marlin/src/module/planner.cpp:
```diff
@@ -1157,17 +1156,16 @@ void Planner::recalculate_trapezoids() {
 
     // Skip sync and page blocks
-    if (!(next->flag & BLOCK_MASK_SYNC) && !IS_PAGE(next)) {
+    if (!next->is_move()) {
```
and here

tombrazier committed:
Fixed inverse logic on is_move()

thinkyhead commented:
Yeah, that's what I meant! The compiler should have realized that.

thinkyhead committed:
fix _ALINE

thinkyhead committed:
update comments

thinkyhead committed:
comments

tombrazier commented:
> Yeah, that's what I meant! The compiler should have realized that.

Compilers can take things so literally.

tombrazier committed:
When printing an arc, tell the planner a safe exit speed after the last block in the queue.

This reduces queue recalculation and also allows faster print speeds.

tombrazier commented:
That's the big optimisation. Now I just need to consolidate all the OPTARG parameters into a struct.

tombrazier committed:
Cleaned up proliferation OPTARGS for the planner

tombrazier commented:
Clean up of OPTARGs complete. Note the `curve_radius` and `safe_exit_speed_sqr` members of `planner_hints_t` could, in principle, be populated for `G5` Bézier curves or even new parameters to G0/G1 to give the same benefits to other paths with many short segments.

tombrazier commented:
I'm done with these changes now.

tombrazier committed:
Don't do below MINIMUM_PLANNER_SPEED

tombrazier commented:
Okay, *now* I'm done.

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/24366

thinkyhead committed:
better with a ClassLike name

thinkyhead committed:
adjust code style

tombrazier commented:
Oh, I have spotted another change. I think the following line needs the safe_exit_speed_sq treatment:

https://github.com/MarlinFirmware/Marlin/blob/8e293f06a02fce899ddb5865dc585e0cefe03dfd/Marlin/src/module/planner.cpp#L2869

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/24366

thinkyhead committed:
misc. cleanup
